plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/JdbcMergeSink.java

```diff
@@ -0,0 +1,322 @@
+                .toArray();
+    }
+
+    private static ConnectorPageSink createUpdateSink(
+            ConnectorSession session,
+            JdbcNamedRelationHandle relation,
+            TupleDomain<ColumnHandle> domain,
+            List<JdbcColumnHandle> primaryKeys,
+            JdbcClient jdbcClient,
+            ConnectorPageSinkId pageSinkId,
+            RemoteQueryModifier remoteQueryModifier,
+            QueryBuilder queryBuilder)
+    {
+        List<JdbcColumnHandle> columns = jdbcClient.getColumns(session, new JdbcTableHandle(relation.getSchemaTableName(), relation.getRemoteTableName(), Optional.empty()));
+
+        ImmutableList.Builder<JdbcAssignmentItem> assignmentItemBuilder = ImmutableList.builder();
+        ImmutableList.Builder<String> columnNamesBuilder = ImmutableList.builder();
+        ImmutableList.Builder<Type> columnTypesBuilder = ImmutableList.builder();
+        QueryParameter dummy = new QueryParameter(BIGINT, Optional.empty());
+        for (JdbcColumnHandle columnHandle : columns) {
+            if (!primaryKeys.contains(columnHandle)) {
+                columnNamesBuilder.add(columnHandle.getColumnName());
+                columnTypesBuilder.add(columnHandle.getColumnType());
+                assignmentItemBuilder.add(new JdbcAssignmentItem(columnHandle, dummy));
+            }
+        }
+        for (JdbcColumnHandle columnHandle : primaryKeys) {
+            columnNamesBuilder.add(columnHandle.getColumnName());
+            columnTypesBuilder.add(columnHandle.getColumnType());
+        }
+
+        return new JdbcPageSink(
+                session,
+                new JdbcOutputTableHandle(
+                        relation.getRemoteTableName(),
+                        columnNamesBuilder.build(),
+                        columnTypesBuilder.build(),
+                        Optional.empty(),
+                        Optional.empty(),
+                        Optional.empty()),
+                jdbcClient,
+                pageSinkId,
+                remoteQueryModifier,
+                updateSqlProvider(session, relation, domain, assignmentItemBuilder.build(), jdbcClient, queryBuilder));
+    }
+
+    private static SinkSqlProvider updateSqlProvider(
+            ConnectorSession session,
+            JdbcNamedRelationHandle relation,
+            TupleDomain<ColumnHandle> domain,
+            List<JdbcAssignmentItem> assignmentItems,
+            JdbcClient jdbcClient,
+            QueryBuilder queryBuilder)
+    {
+        try (Connection connection = jdbcClient.getConnection(session)) {
+            return (_, _, _) -> queryBuilder.prepareUpdateQuery(
+                            jdbcClient,
+                            session,
+                            connection,
+                            relation,
+                            domain,
+                            Optional.empty(),
+                            assignmentItems)
+                    .query();
```

Comment:
I don't think you should use simple query like that. I think you should use parameters too do a column mapping. Now you have tight coupling between this class and query builder. And coupling is not on contract level, but on implementation details.

In `io.trino.plugin.jdbc.JdbcClient#buildInsertSql` is a bit better as order matches the order of passed arguments, but assuming that assignments are handled before domain is a bit risky long term.

Reply:
I am not sure I understand. Do you mean it is better to not return query and instead should return a PreparedQuery here, then use it somehow maybe like `DefaultQueryBuilder#prepareStatement` to handle the mapping?

Reply:
We can not get the value for parameters on constructor when build the `updateSink` or `deleteSink`, and if we want to use parameters then probably not able to reuse `JdbcPageSink`, maybe create a class called JdbcUpdatePageSink?

Reply:
Or do you mean I need to build the query directly, not use the `DefaultQueryBuilder`?

Reply:
> Do you mean it is better to not return query and instead should return a PreparedQuery here, then use it somehow maybe like DefaultQueryBuilder#prepareStatement to handle the mapping?

Yes. `QueryBuilder` is designed to handle single list of parameters. It works well for `SELECT` queries. However here we are using batch of queries. I wonder what it would take to change `QueryBuilder` to be able to support batch too. Then maybe `QueryParameter` should not have a value, but some kind of a "placeholder" which could be then filled for `SELECT` kind of queries or batches.

I guess we are going out of the scope of this PR. Maybe we could do it as follow up.